File: app/src/main/java/org/neidhardt/dynamicsoundboard/dialog/soundlayouts/SoundLayoutSettingsDialog.java
package org.neidhardt.dynamicsoundboard.dialog.soundlayouts;

import android.app.FragmentManager;
import android.os.Bundle;
import org.neidhardt.dynamicsoundboard.NavigationDrawerFragment;
import org.neidhardt.dynamicsoundboard.R;
import org.neidhardt.dynamicsoundboard.soundlayouts.SoundLayoutsManager;

/**
 * Created by eric.neidhardt on 12.03.2015.
 */
public class SoundLayoutSettingsDialog extends SoundLayoutDialog
{
	private static final String TAG = SoundLayoutSettingsDialog.class.getName();

	private static final String KEY_DATABASE_ID = "org.neidhardt.dynamicsoundboard.dialog.SoundLayoutSettingsDialog.databaseId";

	private String databaseId;

	public static void showInstance(FragmentManager manager, String databaseId)
	{
		SoundLayoutSettingsDialog dialog = new SoundLayoutSettingsDialog();

		Bundle args = new Bundle();
		args.putString(KEY_DATABASE_ID, databaseId);
		dialog.setArguments(args);

		dialog.show(manager, TAG);
	}

	@Override
	public void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);

		Bundle args = this.getArguments();
		if (args != null)
			this.databaseId = args.getString(KEY_DATABASE_ID);
	}

	@Override
	protected int getLayoutId()
	{
		return R.layout.dialog_sound_layout_settings;
	}

	@Override
	protected String getHintForName()
	{
		return SoundLayoutsManager.getInstance().getSoundLayoutById(this.databaseId).getLabel();
	}

	@Override
	protected void deliverResult()
	{
		String name = super.soundLayoutName.getDisplayedText();
		SoundLayoutsManager.getInstance().updateSoundLayoutById(this.databaseId, name);

		NavigationDrawerFragment fragment = this.getNavigationDrawerFragment();
		if (fragment != null)
			fragment.triggerSoundLayoutUpdate();
	}
}

File: app/src/main/kotlin/org/neidhardt/dynamicsoundboard/soundlayoutmanagement/views/SoundLayoutSettingsDialog.kt
package org.neidhardt.dynamicsoundboard.soundlayoutmanagement.views

import android.os.Bundle
import android.support.v4.app.FragmentManager
import android.widget.Toast
import org.greenrobot.eventbus.EventBus
import org.neidhardt.dynamicsoundboard.R
import org.neidhardt.dynamicsoundboard.SoundboardApplication
import org.neidhardt.dynamicsoundboard.soundlayoutmanagement.events.SoundLayoutRenamedEvent
import org.neidhardt.dynamicsoundboard.soundlayoutmanagement.model.findById
import rx.android.schedulers.AndroidSchedulers
import rx.subscriptions.CompositeSubscription

/**
 * File created by eric.neidhardt on 12.03.2015.
 */
class SoundLayoutSettingsDialog : SoundLayoutDialog() {

	private val soundLayoutsManager = SoundboardApplication.soundLayoutManager
	private val subscriptions = CompositeSubscription()

	private var databaseId: String? = null

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)

		val args = this.arguments
		if (args != null)
			this.databaseId = args.getString(KEY_DATABASE_ID)
	}

	override fun onDestroy() {
		super.onDestroy()
		this.subscriptions.unsubscribe()
	}

	override fun deliverResult() {
		if (this.databaseId != null) {
			var name = super.soundLayoutName?.text.toString()
			if (name.isEmpty())
				name = this.getHintForName()

			val existingItem = this.soundLayoutsManager.soundLayouts.findById(this.databaseId!!) ?: return

			this.subscriptions.add(
					this.soundLayoutsManager.updateSoundLayout {
						existingItem.apply { this.label = name }
					}
					.observeOn(AndroidSchedulers.mainThread())
					.subscribe( { updatedItem ->
						EventBus.getDefault().post(SoundLayoutRenamedEvent(updatedItem))
					}, { error ->
						Toast.makeText(this.activity, R.string.sound_layouts_toast_change_error, Toast.LENGTH_SHORT).show()
						this.dismiss()
					}, {
						this.dismiss()
					})
			)
		}
	}

	override fun getPositiveButtonId(): Int = R.string.dialog_rename

	override fun getLayoutId(): Int = R.layout.dialog_sound_layout_settings

	override fun getHintForName(): String = this.soundLayoutsManager.soundLayouts.findById(this.databaseId!!)?.label ?: ""

	override fun getTitleId(): Int = R.string.dialog_sound_layout_settings_title

	companion object {
		private val TAG = SoundLayoutSettingsDialog::class.java.name

		private val KEY_DATABASE_ID = "org.neidhardt.dynamicsoundboard.dialog.SoundLayoutSettingsDialog.databaseId"

		fun showInstance(manager: FragmentManager, databaseId: String) {
			val dialog = SoundLayoutSettingsDialog()

			val args = Bundle()
			args.putString(KEY_DATABASE_ID, databaseId)
			dialog.arguments = args

			dialog.show(manager, TAG)
		}
	}
}
